src/dhtproto/node/neo/request/Update.d

```diff
@@ -0,0 +1,329 @@
+        Copyright (c) 2018 sociomantic labs GmbH. All rights reserved.
+
+    License:
+        Boost Software License Version 1.0. See LICENSE.txt for details.
+
+*******************************************************************************/
+
+module dhtproto.node.neo.request.Update;
+
+import swarm.neo.node.IRequestHandler;
+
+/*******************************************************************************
+
+    v0 Update request protocol.
+
+*******************************************************************************/
+
+public abstract class UpdateProtocol_v0 : IRequestHandler
+{
+    import swarm.neo.node.RequestOnConn;
+    import dhtproto.common.Update;
+    import dhtproto.node.neo.request.core.Mixins;
+
+    import ocean.transition;
+    import ocean.core.array.Mutation : copy;
+    import ocean.io.digest.Fnv1;
+
+    /// Mixin the initialiser and the connection and resources members.
+    mixin IRequestHandlerRequestCore!();
+
+    /// Slice of acquired buffer into which the initial payload received from
+    /// the client is copied.
+    private Const!(void)[] init_payload;
+
+    /***************************************************************************
+
+        Called by the connection handler immediately after the request code and
+        version have been parsed from a message received over the connection.
+        Allows the request handler to process the remainder of the incoming
+        message, before the connection handler sends the supported code back to
+        the client.
+
+        Note: the initial payload is a slice of the connection's read buffer.
+        This means that when the request-on-conn fiber suspends, the contents of
+        the buffer (hence the slice) may change. It is thus *absolutely
+        essential* that this method does not suspend the fiber. (This precludes
+        all I/O operations on the connection.)
+
+        Params:
+            init_payload = initial message payload read from the connection
+
+    ***************************************************************************/
+
+    public void preSupportedCodeSent ( Const!(void)[] init_payload )
+    {
+        auto ed = this.connection.event_dispatcher();
+        auto buf = this.resources.getVoidBuffer();
+        (*buf).copy(init_payload);
+        this.init_payload = *buf;
+    }
+
+    /***************************************************************************
+
+        Called by the connection handler after the supported code has been sent
+        back to the client.
+
+    ***************************************************************************/
+
+    public void postSupportedCodeSent ( )
+    {
+        auto ed = this.connection.event_dispatcher();
+
+        auto message =
+            *ed.message_parser.getValue!(MessageType)(this.init_payload);
+        auto channel = ed.message_parser.getArray!(char)(this.init_payload);
+        auto key = *ed.message_parser.getValue!(hash_t)(this.init_payload);
```

Comment:
Suggest using `ed.message_parser.parseBody`.

Reply:
The advantage of `parseBody` is to detect more message protocol errors by ensuring the message has only the expected fields and not additional data. If I understand the protocol for this request correctly, you could use it like this:

```d
auto message = *ed.message_parser.getValue!(MessageType)(this.init_payload);
// ...
with ( MessageType ) switch ( message )
{
    // Normal sequence: read, wait for response, update.
    case GetRecord:
        cstring key;
        ed.message_parser.parseBody(this.init_payload, key);
        response = this.getWaitUpdate(channel, key);
        break;

    // Either an updated record transferred from another node or a
    // new record that did not exist in the DHT before. Just put it.
    case UpdateRecord:
        cstring key;
        hash_t original_hash;
        const!(void)[] new_value;
        // Read original hash and new value here rathern than in `update`
        ed.message_parser.parseBody(this.init_payload, key, original_hash, new_value);
        response = this.update(channel, key, original_hash, new_value);
        break;

    default:
        ed.shutdownWithProtocolError("Unexpected message from client");
}
```

Reply:
Ping on `parseBody`

Reply:
It can't be used here, as the payload is passed on to `update`, where more fields are read.

Reply:
All clear, thanks!